zio-http/shared/src/main/scala/zio/http/endpoint/openapi/JsonSchema.scala

```diff
@@ -1248,23 +1285,79 @@ object JsonSchema {
     def required(required: Chunk[java.lang.String]): Object =
       this.copy(required = required)
 
+    private def reconcileTwoOptions[T](left: Option[T], right: Option[T], noneIfOne: Boolean)(
+      combine: (T, T) => Option[T],
+    ): Option[T] = {
+      if (noneIfOne) left.fold(Option.empty[T])(l => right.fold(Option.empty[T])(r => combine(l, r)))
+      else left.fold(right)(l => right.fold(left)(r => combine(l, r)))
+    }
+
+    private def someWhenEq[T](l: T, r: T): Option[T] =
+      if (l == r) Some(l) else None
+
+    private def combinePatterns(lPattern: Pattern, rPattern: Pattern): Some[Pattern] =
+      if (lPattern == rPattern) Some(lPattern)
+      else {
+        // validate both patterns match using lookahead assertions
+        Some(Pattern("(?=" + lPattern + ")(?=" + rPattern + ")"))
+      }
+
+    private def wrap[T](f: (T, T) => T): (T, T) => Option[T] = (l, r) => Some(f(l, r))
+
     private def combineAdditionalProperties(
       left: Either[Boolean, JsonSchema],
       right: Either[Boolean, JsonSchema],
     ): Either[Boolean, JsonSchema] =
       (left, right) match {
-        case (Left(false), _)            => Left(false)
-        case (_, Left(_))                => left
-        case (Left(true), _)             => right
-        case (Right(left), Right(right)) =>
-          Right(AllOfSchema(Chunk(left, right)))
+        case (Left(false), _)                 => Left(false)
+        case (_, Left(_))                     => left
+        case (Left(true), _)                  => right
+        case (Right(lSchema), Right(rSchema)) =>
+          val (leftKey, lAnnotations)  = Object.extractKeySchemaFromAnnotations(lSchema)
+          val (rightKey, rAnnotations) = Object.extractKeySchemaFromAnnotations(rSchema)
+
+          val keySchema = reconcileTwoOptions(leftKey, rightKey, noneIfOne = true) {
+            // TODO: what happens in case of annotated key schemas?
+            //       Is it possible?
+            //       How should we combine if so?
+            case (JsonSchema.String(lfmt, lptn, lmxl, lmnl), JsonSchema.String(rfmt, rptn, rmxl, rmnl)) =>
+              Some(
+                JsonSchema.String(
+                  format = reconcileTwoOptions(lfmt, rfmt, noneIfOne = true)(someWhenEq),
```

Comment:
Not sure, if specialized methods with some duplicated code are not easier to maintain.

Reply:
you mean instead of having the flag?
OK, makes sense.